### PR TITLE
WSDL Test attempt no. 2

### PR DESCRIPTION
**Pull request by ppalaga**

### CODE REVIEW

File: cxf-soap/src/main/java/org/acme/cxf/soap/service/ContactService.java [hunk 1]
```diff
@@ -16,13 +16,17 @@
...
+@WebService(name = "ContactService", serviceName = "ContactService", targetNamespace = ContactService.TARGET_NS)
 public interface ContactService {
+    public static final String TARGET_NS = "http://camel.apache.org/test/ContactService";
 
     void addContact(Contact contact);
 
     Contact getContact(String name) throws NoSuchContactException;
 
-    GetContactsResponse getContacts();
+    Contacts getContacts();
```
Comment (llowinge) [inline cxf-soap/src/main/java/org/acme/cxf/soap/service/ContactService.java:29]: Why we cannot return `List<Contact>` ?

File: cxf-soap/src/test/java/org/acme/cxf/soap/PojoClientTest.java [hunk 2]
```diff
@@ -1,26 +1,33 @@
...
-        String URL = getServerUrl() + "/cxf/services/contact";
-
-        ClientProxyFactoryBean factory = new ClientProxyFactoryBean();
-        factory.setServiceClass(ContactService.class);
-        factory.setAddress(URL);
-
-        return (ContactService) factory.create();
+        try {
+            final URL serviceUrl = new URL(getServerUrl() + "/cxf/services/contact?wsdl");
+            final QName qName = new QName(ContactService.TARGET_NS, ContactService.class.getSimpleName());
```
Comment (llowinge) [inline cxf-soap/src/test/java/org/acme/cxf/soap/PojoClientTest.java:25]: Why is this change needed ?

Comment (ppalaga) [reply cxf-soap/src/test/java/org/acme/cxf/soap/PojoClientTest.java:25]: Because otherwise the client seems to miss the info from the WSDL served via `/cxf/services/contact?wsdl` and those marshaling exceptions would occur. Not sure it does/should/would work also without the explicit namespace.

File: cxf-soap/src/main/java/org/acme/cxf/soap/service/ContactService.java (hunk 1 above)
Comment (ppalaga) [reply cxf-soap/src/main/java/org/acme/cxf/soap/service/ContactService.java:29]: IIRC, Lists do not work well in general in JAX-WS without these wrappers. Arrays seem to be better supported. But when returning `Contact[]` from getContacts(), a bug pops up caused by the fact that Camel or CXF replaces a single element array by the element. Not sure how to workaround it. Maybe some binding config or transformers.

File: cxf-soap/src/test/java/org/acme/cxf/soap/PojoClientTest.java (hunk 2 above)
Comment (llowinge) [reply cxf-soap/src/test/java/org/acme/cxf/soap/PojoClientTest.java:25]: In spring-boot it works (https://github.com/jboss-fuse/camel-spring-boot-examples/blob/camel-spring-boot-examples-3.20.1-branch/soap-cxf/src/test/java/sample/camel/PojoClientTest.java#L42) probably because https://github.com/jboss-fuse/camel-spring-boot-examples/blob/camel-spring-boot-examples-3.20.1-branch/soap-cxf/src/main/java/sample/camel/service/ContactService.java#L19 doesn't have the `@WebService` annotation ?

Comment (llowinge) [reply cxf-soap/src/test/java/org/acme/cxf/soap/PojoClientTest.java:25]: @ffang Could you please check - if the web service has `@WebService` annotation and the `ContactService` created from `ClientProxyFactoryBean` (without QName specifying targetNS) should work ?

Comment (ffang) [reply cxf-soap/src/test/java/org/acme/cxf/soap/PojoClientTest.java:25]: Hi @llowinge ,

The difference comes from @WebService annotation indeed.
The class ClientProxyFactoryBean is used for Simple frontend(simple frontend means the SEI/Service Endpoint Interface is a POJO, no JAXWS annotation) to create a client proxy. As ContactService now has JAXWS annotation, you should use JaxWsProxyFactoryBean which is from jaxws frontend to create the client proxy.

Both ClientProxyFactoryBean and JaxWsProxyFactoryBean are CXF API to create client proxy, they can play the same role as using JAXWS API(The way Peter is using) to create client proxy

Freeman

Comment (llowinge) [reply cxf-soap/src/test/java/org/acme/cxf/soap/PojoClientTest.java:25]: @ffang Thank you so much for clarification. It makes much more sense now to me. I've somehow missed the "Simple frontend" approach.

File: cxf-soap/src/main/java/org/acme/cxf/soap/service/ContactService.java (hunk 1 above)
Comment (llowinge) [reply cxf-soap/src/main/java/org/acme/cxf/soap/service/ContactService.java:29]: Ok thank you.
